src/main/scala/org/clulab/wm/eidos/EidosSystem.scala

```diff
@@ -28,140 +24,153 @@ class EidosSystem(val components: EidosComponents) {
+      val mostCompleteEvents = components.actions.keepMostCompleteEvents(extractedEvents, State(extractedEvents))
 
       mostCompleteEvents
     }
 
     require(doc.text.isDefined)
-    // Prepare the initial state -- if you are using the entity finder then it contains the found entities,
-    // else it is empty
+    // Prepare the initial state.  If you are using the entity finder, then it contains the found entities;
+    // otherwise, it is empty.
     val initialState = components.entityFinders.foldLeft(new State()) { (state, entityFinder) =>
       val mentions = entityFinder.find(doc, state)
+
       state.updated(mentions)
     }
 
-    // Run the main extraction engine, pre-populated with the initial state
-    val events = extractEventsFrom(doc, initialState).distinct
+    // Run the main extraction engine, pre-populated with the initial state.
+    val events = extractMentions(initialState).distinct
     // Note -- in main pipeline we filter to only CAG relevant after this method.  Since the filtering happens at the
-    // next stage, currently all mentions make it to the webapp, even ones that we filter out for the CAG exports
+    // next stage, currently all mentions make it to the webapp, even ones that we filter out for the CAG exports.
     //val cagRelevant = keepCAGRelevant(events)
 
     events
   }
 
-  def postProcess(annotatedDocument: AnnotatedDocument): AnnotatedDocument = {
-    val lastAnnotatedDocument = postProcessors.foldLeft(annotatedDocument) { (nextAnnotatedDocument, postProcessor) =>
-      postProcessor.process(nextAnnotatedDocument)
+  def preProcess(preProcessors: Seq[PreProcessing], odinMentions: Seq[Mention]): Seq[Mention] = {
+    val lastOdinMentions = preProcessors.foldLeft(odinMentions) { (prevOdinMentions, preProcessor) =>
+      val nextOdinMentions = preProcessor(prevOdinMentions)
+
+      nextOdinMentions // debug here
+    }
+
+    lastOdinMentions
+  }
+
+  def postProcess(postProcessors: Seq[PostProcessing], annotatedDocument: AnnotatedDocument): AnnotatedDocument = {
+    val lastAnnotatedDocument = postProcessors.foldLeft(annotatedDocument) { (prevAnnotatedDocument, postProcessor) =>
+      val nextAnnotatedDocument = postProcessor.process(prevAnnotatedDocument)
+
+      nextAnnotatedDocument // debug here
     }
 
     lastAnnotatedDocument
   }
 
+  // Grounding is the first PostProcessing step(s) and it is pre-configured in Eidos.  Other things can take
+  // the resulting AnnotatedDocument and post-process it further.  They are not yet integrated into Eidos.
+  protected def mkPostProcessors(options: EidosSystem.Options): Seq[PostProcessing] = Seq(
+    components.ontologyHandler
+  )
+
+  protected def mkPreProcessors(options: EidosSystem.Options): Seq[PreProcessing] = {
+    Seq(
+      (odinMentions: Seq[Mention]) => { components.conceptExpander.expand(odinMentions) },
+      (odinMentions: Seq[Mention]) => { nestedArgumentExpander.traverse(odinMentions) },
```

Comment:
IDK what the consequences of expanding everything after running the grammars are.  If an argument is expanded in one "epoch/priority" of the grammar, then as a result other rules may apply differently.
Maybe this is fine, maybe it's not but can be fixed with tweaks to the grammars, etc.
Did you notice a difference in extracted event quantity?

Reply:
Maybe I missed something.  The functionality is not supposed to change.  The old code was
```scala
val afterExpandingConcepts = components.conceptExpander.expand(odinMentions)
val mentionsAndNestedArgs =  traverse(afterExpandingConcepts, Seq.empty, Set.empty)
val cagRelevant =...
```
Maybe I've confused things with a misnamed nestedArgumentExpander.  I didn't have a good name for that.  Suggestions welcome.

Reply:
OH! you're right, sorry I think I was thinking about an older workflow that is no longer applied! Disregard!!!